bin/propolis-server/src/lib/vm/state_driver.rs

```diff
@@ -520,20 +534,167 @@ impl StateDriver {
+        // task, which can decide whether and how to buffer incoming requests.
+        drop(objects);
+
+        loop {
+            let event = tokio::select! {
+                // If the VM successfully starts, return immediately and let
+                // the caller process any events that may happen to be on the
+                // queue.
+                biased;
+
+                res = &mut block_backend_fut => {
+                    if res.is_ok() {
+                        let objects = &self.objects;
+                        objects.lock_exclusive().await.resume_vcpus();
+                        self.publish_steady_state(InstanceState::Running);
+                        info!(&self.log, "VM successfully started");
+                    }
+
+                    return res;
+                }
+
+                dequeued = self.input_queue.wait_for_next_event() => {
+                    dequeued
+                }
+            };
+
+            // The VM's vCPUs haven't been started yet, so there should be no
+            // way for the VM to produce any guest events to handle.
+            let InputQueueEvent::ExternalRequest(req) = event else {
+                unreachable!("can't get guest events before the VM starts");
+            };
+
+            // Handle requests to reconfigure one of the existing Crucible
+            // volumes inline, but buffer other requests so that they can be
+            // handled after the VM has finished starting.
+            //
+            // Buffering some requests and servicing others can theoretically
+            // change the order in which those requests are retired. That's not
+            // a problem here because the request queue will stop accepting new
+            // VCR change requests once a request to stop has been queued.
+            match req {
+                ExternalRequest::Stop => {
+                    info!(
+                        &self.log,
+                        "got request to stop while still starting"
+                    );
+
+                    // Buffer this request to stop so that it can be processed
+                    // as part of the main event loop once the VM finishes
+                    // starting.
+                    //
+                    // It is possible, at least in theory, to drop the block
+                    // backend startup future (or cooperatively cancel it) and
+                    // then pause/halt the VM immediately. This requires block
+                    // backends' startup operations to be cancel-safe (in the
+                    // sense that it must be possible to pause and halt a
+                    // backend whose startup sequence was interrupted). Since
+                    // block backends aren't currently required to guarantee
+                    // this, just buffer the stop request.
+                    self.input_queue.buffer_external_request(req);
```

Comment:
I'm worried about the interaction of this line and `self.input_queue.wait_for_next_event()` in the `select!` above.

Specifically, `wait_for_next_event()` will return events from the `buffered_external` queue, so I'm suspicious that if we send `ExternalRequest::Stop` while the machine is starting, we'll end up in a spin-loop of removing then re-adding the request to the `buffered_external` quee.

Reply:
Great catch here--the new test case I added for this in bfb8b07 fails due to this problem. That commit should also fix the issue (by waiting to self-request a stop until the attempt to start has actually finished).